### DISCUSSION TITLE
The Road to `next-mdx-remote` v5 and MDX v3

### DISCUSSION THREAD

**Discussion by dstaley** [hashicorp/next-mdx-remote#437] [Announcements]
**Update 2024-05-22**: `next-mdx-remote@5.0.0` is now available! :tada:

---

**Update 2024-03-13**: The first canary releases of `next-mdx-remote` are now available for testing! [More details in the base PR](https://github.com/hashicorp/next-mdx-remote/pull/448).

---

Hello everyone! Apologies for the radio silence for the past couple of months. I wanted to take a moment and provide a bit of background on where HashiCorp currently is in relation to `next-mdx-remote`, and give y'all an overview of where we'd like to go with it in the near future.

## Background

HashiCorp began adopting MDX in early 2019, right around the release of MDX v1.0.0. Since then, documentation writers at HashiCorp have amassed over 40,000 MDX docs that serve as the content for [HashiCorp Developer](https://developer.hashicorp.com). When MDX v2 was released in 2022 with significant breaking changes, we were unable to carve out enough engineering time to perform the upgrade, but we knew that we'd eventually migrate, so we took the time to release `next-mdx-remote` v4.0.0, which updated to MDX v2. It's now been two years since that release, and we're just beginning down the path of adopting MDX v2 compatibility within our infrastructure. Since our products still use `next-mdx-remote` v3, it was difficult to justify allocating engineering efforts towards the maintenance of v4. This has resulted in issues piling up un-triaged, and pull requests that have been unanswered.

## Moving Forward

It's our hope that, in the near future, all 40,000+ of our docs will be compatible with MDX v2/v3, and that we can update to the latest version of `next-mdx-remote`. In support of that, we're going to allocate engineering time to getting things tidied up, and releasing v5 of `next-mdx-remote`, which will contain support for MDX v3.

Currently, this is our plan:

1. Get CI tests passing again. (Fixed in #435)
1. Add support for our standard package publishing workflow, which will enable per-commit canary releases. (Added in #446)
1. Create a `v5` branch, update to MDX v3, and fix any issues that arise as a result of the update (including any necessary updates to tooling like Jest). (#448)
1. Publish an initial canary release with v3 support, followed by fixing bugs in the core functionality of `next-mdx-remote`.
1. Publish `next-mdx-remote` v5.

It should be noted that `next-mdx-remote` was designed with the functionality of MDX v1 in mind, and certain v2/v3 features might not fit neatly into the existing API of v4 (which we aim to maintain with v5 where possible). Because of this, we will be focused on MDX v3 compatibility and bug fixes for v5, and reserve new features for v6. Following the initial release of v5 we will look into replacing some of our tooling (like Rollup and Jest) with the goal of eliminating the need to manually update TypeScript definition files and improving compatibility with various module formats.

## How You Can Help

Since HashiCorp doesn't have a production application running `next-mdx-remote` v4, we'd greatly appreciate it if you could share examples of open source projects using v4. This will give us a test bed for future work, and help ensure that we're shipping a version that is compatible with existing codebases. To share an example, please leave a comment in [this discussion](https://github.com/hashicorp/next-mdx-remote/discussions/439).

Also, if you're maintaining a fork of `next-mdx-remote` or have an alternative library that users of `next-mdx-remote` might be interested in, please leave a comment in [this discussion](https://github.com/hashicorp/next-mdx-remote/discussions/438) so that we can add your library to a list of alternatives in the README. We are especially interested in sharing links to alternatives that incorporate features not found in `next-mdx-remote`, and that may be a better fit for users requiring more advanced functionality.

Finally, I want to thank everyone for their patience and understanding, and give a special shout out to those that have shared their forks with others. One of my favorite aspects of open source is seeing people helping one another, and seeing forks that arose out of a specific need grow and develop. Even though we're going to be putting some work into `next-mdx-remote`, please don't let that stop you from continuing to work on your alternatives! We want nothing more than to see the MDX ecosystem grow and flourish.

**Comment by mryechkin** [comment]
Thank you for the update. I'll admit I was one of those people that was starting to get worried about the future of this project - glad to hear that it's not going anywhere! 🥳

**Comment by joshwcomeau** [comment]
Hi there! Glad to hear work on this project is continuing 😄.

I have a [React course](https://joyofreact.com/), and the final project in that course uses next-mdx-remote v4. It's not quite a sprawling real-world example, but it does use next-mdx-remote v4, and feature blog posts with dynamic content.

Github: https://github.com/joy-of-react/project-blog (check out the `solution` branch for the full thing)
Live: https://www.final-project.blog/

**Comment by wdavidw** [comment]
The official [TDP (Trunk Data Platform)](https://www.trunkdataplatform.io/) website is running with version 4.4.1 and it is fully [open source](https://github.com/TOSIT-IO/tdp-website). Note, it is affected with issue #403.

**Comment by talatkuyuk** [comment]
Hi @dstaley,

Thank you for sharing your thoughts and plans regarding the package feature. I'm a huge fan of the unified, remark, mdx ecosystem and have been actively contributing in my spare time, which I thoroughly enjoy.

I've also been actively commenting on the issues within `next-mdx-remote` thus far.

In early January 2024, I decided to fork `next-mdx-remote` and update it, with the intention of making a PR. However, it seems that PRs haven't been considered for quite some time.

As I made the changes, a brand-new package emerged. I leveraged the beneficial features of `next-mdx-remote`, propelled it forward. It took two months to develop the new package, including tests and documentation.  I aligned it with the mindset of **`@mdx-js/mdx` version 3**.  I'm truly grateful to @wooorm and his associates for developing the unified ecosystem.

After reading your comment,  I felt encouraged to share the new package with the community.

The **[`next-mdx-remote-client`](https://github.com/ipikuka/next-mdx-remote-client)** serves as a viable alternative to `next-mdx-remote` for the time being, offering a solution until `next-mdx-remote` receives updates in the future. I appended the term "**client**" to the package name to emphasize the exported components for both server and client, tailored for the "app" router and the "pages" router respectively.

I published the **`next-mdx-remote-client`** on npm as [version 1.0.0](https://www.npmjs.com/package/next-mdx-remote-client). 

I implemented a demo app as well --> https://github.com/talatkuyuk/demo-next-mdx-remote-client

It primarily supports **imports** and **exports** in MDX in the "app" router and **only exports** in the "pages" router. Additionally, it provides options to disable exports and imports, along with extra features requested by the community. I've made an effort to address all the reported issues in order to resolve any problems. It works with latest `nextjs` ^13.5.6 and ^14.1.0 versions (tested).

I'm eager to hear the community's feedback. Thank you.

**Comment by dominik-sfl** [comment]
Any news on when v5 might be coming?

**Reply by dstaley** [reply]
See [this issue](https://github.com/hashicorp/next-mdx-remote/issues/457) for more details.
